crates/subspace-networking/src/node_runner.rs

```diff
@@ -1172,32 +1163,15 @@ where
         if let AutonatEvent::StatusChanged { old, new } = event {
             info!(?old, ?new, "Public address status changed.");
 
-            if matches!(self.kademlia_mode, KademliaMode::Dynamic { .. }) {
-                let mode = match &new {
-                    NatStatus::Public(address) => {
-                        if is_global_address_or_dns(address)
-                            || self.allow_non_global_addresses_in_dht
-                        {
-                            Mode::Server
-                        } else {
-                            debug!(
-                                ?old,
-                                ?new,
-                                ?address,
-                                non_global_addresses=%self.allow_non_global_addresses_in_dht,
-                                "Kademlia mode wasn't set."
-                            );
-                            Mode::Client
-                        }
-                    }
-                    NatStatus::Private => Mode::Client,
-                    NatStatus::Unknown => Mode::Client,
-                };
-
-                self.swarm.behaviour_mut().kademlia.set_mode(Some(mode));
+            // TODO: Remove block once https://github.com/libp2p/rust-libp2p/issues/4863 is resolved
+            if let (NatStatus::Public(old_address), NatStatus::Private) = (old, new) {
+                self.swarm.remove_external_address(&old_address);
+                // Trigger potential mode change manually
+                self.swarm.behaviour_mut().kademlia.set_mode(None);
```

Comment:
That was a simpler workaround rather than proper fix, with multiple addresses properly supported in this PR (to the best of my knowledge) this should not longer be necessary, it should work properly.